### PR TITLE
Lysenko/ocd 1427 rollout actions

### PR DESCRIPTION
**Pull request by konstantin-lysenko-netapp**
[OCD-1427](https://spotinst.atlassian.net/browse/OCD-1427?atlOrigin=eyJpIjoiZTgzM2UyMDMxNmMwNDYyMzk1MzgzZDE2MDhjNGU2NDAiLCJwIjoiaiJ9)_

# Checklist:
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have validated all the requirements in the Jira task were answered
- [x] I have all neccessary approvals for the design/mini design of this task

### CODE REVIEW

File: cmd/rollout.go [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+	// Cobra supports Persistent Flags which will work for this command
+	// and all subcommands, e.g.:
+	// explainCmd.PersistentFlags().String("foo", "", "A help for foo")
+
+	// Cobra supports local flags which will only run when this command
+	// is called directly, e.g.:
+	// explainCmd.Flags().BoolP("toggle", "t", false, "Help message for toggle")
+}
+
+func runRolloutActionCmd(action string, args []string, actionPastForm string) {
```
Comment (caduri) [inline cmd/rollout.go:36]: change to `runRolloutAction`

File: pkg/oceancd/api.go [hunk 2]
```diff
@@ -195,3 +195,29 @@ func unmarshalEntityResponse(response []byte) ([]interface{}, error) {
 	return b, nil
 
 }
+
+func RolloutPut(rolloutID string, body map[string]string) error {
```
Comment (caduri) [inline pkg/oceancd/api.go:199]: change to `sendRolloutAction`

**Review by caduri — CHANGES_REQUESTED** [review]

File: cmd/abort.go [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+// abortCmd represents the abort command
+var (
+	abortDescription = "The rollout will be terminated and the previous version (i.e., Stable) will be restored"
+	abortCmd         = &cobra.Command{
+		Use:   oceancd.AbortAction + " ROLLOUT_ID",
+		Short: abortDescription,
+		Long:  abortDescription,
+		Args: func(cmd *cobra.Command, args []string) error {
+			return validateRolloutActionArgs(cmd, args)
+		},
```
[the remark this answers is not in the document]
Comment (konstantin-lysenko-netapp) [reply cmd/abort.go:17]: could we do it globally using the root command or on initialization?

File: pkg/oceancd/api.go (hunk 2 above)
Comment (konstantin-lysenko-netapp) [reply pkg/oceancd/api.go]: done

File: cmd/rollout.go (hunk 1 above)
Comment (konstantin-lysenko-netapp) [reply cmd/rollout.go]: done

File: cmd/rollout.go [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+	} else {
+		rolloutID := args[0]
+		if strings.HasPrefix(rolloutID, "rol-") == false {
+			fmt.Println(`Rollout ID must have the "rol-" prefix.`)
+			return errors.New("error: Invalid Rollout ID")
+		}
+
+		if false == regexp.MustCompile(`^[a-zA-Z\d]*$`).MatchString(strings.TrimPrefix(rolloutID, "rol-")) {
+			fmt.Println(`Rollout ID must only contain letters and digits after the "rol-" prefix.`)
+			return errors.New("error: Invalid Rollout ID")
```
[the remark this answers is not in the document]
Comment (konstantin-lysenko-netapp) [reply cmd/rollout.go]: done

File: cmd/rollout.go [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+		return errors.New("error: Too many arguments")
+	} else {
+		rolloutID := args[0]
+		if strings.HasPrefix(rolloutID, "rol-") == false {
+			fmt.Println(`Rollout ID must have the "rol-" prefix.`)
+			return errors.New("error: Invalid Rollout ID")
+		}
+
+		if false == regexp.MustCompile(`^[a-zA-Z\d]*$`).MatchString(strings.TrimPrefix(rolloutID, "rol-")) {
+			fmt.Println(`Rollout ID must only contain letters and digits after the "rol-" prefix.`)
```
[the remark this answers is not in the document]
Comment (konstantin-lysenko-netapp) [reply cmd/rollout.go]: done

File: cmd/rollout.go [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+		fmt.Println("You must specify a rollout ID.")
+		return errors.New("error: Rollout ID not specified")
+	} else if len(args) > 1 {
+		fmt.Println("You can only specify one rollout ID.")
+		return errors.New("error: Too many arguments")
+	} else {
+		rolloutID := args[0]
+		if strings.HasPrefix(rolloutID, "rol-") == false {
+			fmt.Println(`Rollout ID must have the "rol-" prefix.`)
+			return errors.New("error: Invalid Rollout ID")
```
[the remark this answers is not in the document]
Comment (konstantin-lysenko-netapp) [reply cmd/rollout.go]: done

File: cmd/abort.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (konstantin-lysenko-netapp) [reply cmd/abort.go:17]: done

[the remark this answers is not in the document]
Comment (caduri) [reply cmd/abort.go:17]: some commands do not need this validation, explain for example
you can add it to rollout command

[the remark this answers is not in the document]
Comment (konstantin-lysenko-netapp) [reply cmd/abort.go:17]: would it be better if we disable it for commands that don't need validation?

[the remark this answers is not in the document]
Comment (caduri) [reply cmd/abort.go:17]: I prefer a positive attitude, tell me what are you doing
its more readable and less error prone

File: cmd/rollout.go [hunk 7]
```diff
@@ -0,0 +1,69 @@
...
+	"fmt"
+	"github.com/spf13/cobra"
+	"regexp"
+	"spot-oceancd-cli/pkg/oceancd"
+	"strings"
+)
+
+// retryCmd represents the retry command
+var (
+	rolloutDescription = "Perform changes on a rollout level"
```
[the remark this answers is not in the document]
Comment (caduri) [reply cmd/rollout.go:17]: IMO it should be
```
This command consists of multiple subcommands which can perform changes on a SpotDeployment rollout
```

File: cmd/retry.go [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+	"fmt"
+	"github.com/spf13/cobra"
+	"spot-oceancd-cli/pkg/oceancd"
+)
+
+// retryCmd represents the retry command
+var (
+	retryDescription      = "Available for the last rolled back SpotDeployment only. With this action you will be able to retry your full rollout"
+	retryShortDescription = "Retry a rollout"
+	retryExample          = fmt.Sprintf(rolloutActionExampleTemplate,
```
Comment (caduri) [inline cmd/retry.go:13]: Add a method in rollout cmd that get the description and action and return the example

File: cmd/rollout.go [hunk 9]
```diff
@@ -0,0 +1,69 @@
...
+	if err != nil {
+		fmt.Printf("Failed to %s the rollout - %s\n", action, err.Error())
+	} else {
+		fmt.Printf("Successfully %s resource %s\n", actionPastForm, rolloutID)
+	}
+}
+
+func validateRolloutActionArgs(_ *cobra.Command, args []string) error {
+	if len(args) < 1 {
+		fmt.Println("You must specify a rollout ID.")
```
[the remark this answers is not in the document]
Comment (caduri) [reply cmd/rollout.go:67]: if its part of 2 words like `rolloutID` change to `rolloutId`

**Review by caduri — CHANGES_REQUESTED** [review]

File: cmd/retry.go (hunk 8 above)
Comment (konstantin-lysenko-netapp) [reply cmd/retry.go:13]: ok

File: cmd/rollout.go [hunk 10]
```diff
@@ -0,0 +1,86 @@
...
+	// explainCmd.Flags().BoolP("toggle", "t", false, "Help message for toggle")
+}
+
+func runRolloutAction(action string, args []string, actionPastForm string) {
+	rolloutID := args[0]
+	actionRequest := map[string]string{"action": action}
+
+	err := oceancd.SendRolloutAction(rolloutID, actionRequest)
+	if err != nil {
+		fmt.Printf("Failed to %s the rollout - %s\n", action, err.Error())
```
[the remark this answers is not in the document]
Comment (konstantin-lysenko-netapp) [reply cmd/rollout.go:59]: done

File: cmd/rollout.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (caduri) [reply cmd/rollout.go:17]: what about this one?

File: cmd/rollout.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (caduri) [reply cmd/rollout.go:67]: and this?

File: cmd/rollout.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (konstantin-lysenko-netapp) [reply cmd/rollout.go:17]: with Julia, we've decided that we prefer the current version

**Review by caduri — APPROVED** [review]

**Pull request merged**